library/alloc/src/vec/mod.rs

```diff
@@ -1541,6 +1541,9 @@ impl<T, A: Allocator> Vec<T, A> {
         }
 
         let len = self.len();
+        if index > len {
```

Comment:
#98755 explains the reasons for the original change, to reduce the number of conditional tests. This PR will increase the number of tests in the common `index < len` case from one to two.

If the `add` is the problem, it could be moved so it's only computed after the `index < len` and `index == len` tests. Something like this:

```rust
        // Space for the new element.
        if len == self.buf.capacity() {
            self.reserve(1);
        }

        unsafe {
            // Do not call `add` until the index is checked, to avoid UB. Also,
            // minimize the number of conditional tests, and avoid a
            // zero-length copy in the `index == len` case. (All this requires
            // the duplicate computation of `p`.)
            if index < len {
                // Shift everything over to make space. (Duplicating the
                // `index`th element into two consecutive places.) Then
                // overwrite the first copy of the `index`th element.
                let p = self.as_mut_ptr().add(index);
                ptr::copy(p, p.add(1), len - index);
                ptr::write(p, element);
            } else if index == len {
                // No elements need shifting. Overwrite the first copy of the
                // `index`th element.
                let p = self.as_mut_ptr().add(index);
                ptr::write(p, element);
            } else {
                assert_failed(index, len);
            }
            self.set_len(len + 1);                                                
        }          
```

Reply:
That would also avoid the unnecessary extra block :)

Reply:
> This PR will increase the number of tests in the common index < len case from one to two, and reinstate the zero-length copy in the index == len case.

I believe I kept the original change where the `ptr::copy` is only called when `index < len`. I thought that was the main reason for the perf speedup from the original PR, however, I can see the extra check in the common path having an impact on perf.

Should I make a new commit to fold the `index > len` check back to the original comparison now or wait for a perf run?

Reply:
Also, we might get a small speedup in the sense that we no longer potentially resize the `Vec` if the index is OOB, though obviously  it would be silly to try to optimize that at the expense of the non-panicking in-bounds case.